### PR TITLE
refactoring: introduce progress monitor

### PR DESCRIPTION
**Pull request by ckrause**
Adds `ProgressMonitor` class, but does not change any functionality. We will use it later to add checkpoints.

### CODE REVIEW

File: src/util.cpp [hunk 1]
```diff
@@ -394,6 +394,44 @@ void AdaptiveScheduler::reset() {
...
+bool ProgressMonitor::isTargetReached() {
+  return getElapsedSeconds() >= target_seconds;
+}
+
+double ProgressMonitor::getProgress() {
+  double progress = static_cast<double>(getElapsedSeconds()) /
+                    static_cast<double>(target_seconds);
+  if (progress > 1.0) {
+    progress = 1.0;
+  }
```
Comment (neoneye) [inline src/util.cpp:423]: If the elapsed seconds return with a negative value, then perhaps `if (progress < 0.0) { progress = 0.0; }`

File: src/util.cpp [hunk 2]
```diff
@@ -394,6 +394,44 @@ void AdaptiveScheduler::reset() {
...
+  return progress;
+}
+
+void ProgressMonitor::writeProgress() {
+  if (!progress_file.empty()) {
```
Comment (neoneye) [inline src/util.cpp:428]: Is `progress_file` ever empty?

File: src/util.cpp [hunk 3]
```diff
@@ -394,6 +394,44 @@ void AdaptiveScheduler::reset() {
...
+ProgressMonitor::ProgressMonitor(int64_t target_seconds,
+                                 const std::string &progress_file)
+    : start_time(std::chrono::steady_clock::now()),
+      target_seconds(target_seconds),
+      progress_file(progress_file) {
+  if (target_seconds <= 0) {
+    throw std::runtime_error("internal error: invalid target duration: " +
+                             std::to_string(target_seconds));
+  }
+}
```
Comment (neoneye) [inline src/util.cpp:406]: Reject target_seconds that are too far into the future, potentially causing undefined behavior or wrap around.

File: src/util.cpp (hunk 2 above)
Comment (ckrause) [reply src/util.cpp]: could be in the future. added error log message for this case

File: src/util.cpp (hunk 3 above)
Comment (ckrause) [reply src/util.cpp:406]: difficult to define a meaningful upper limit

**Pull request merged**